lib/ansible/modules/system/parted.py

```diff
@@ -0,0 +1,630 @@
+       e":"partition-table-type":"model-name";
+     - If the first line was either CYL or CHS, the next line will contain
+       information on no. of cylinders, heads, sectors and cylinder size.
+     - Partition information begins from the next line. This is of the format:
+       (for BYT)
+       "number":"begin":"end":"size":"filesystem-type":"partition-name":"flags-s
+       et";
+       (for CHS/CYL)
+       "number":"begin":"end":"filesystem-type":"partition-name":"flags-set";
+    """
+    lines = output.split('\n')
+    generic_params = lines[1].rstrip(';').split(':')
+    generic = {
+        'dev':   generic_params[0],
+        'size':  parse_unit(generic_params[1], unit),
+        'table': generic_params[5],
+        'model': generic_params[6],
+        'logical_block':  int(generic_params[3]),
+        'physical_block': int(generic_params[4])
+    }
+
+    parts = []
+    for line in lines[2:]:
+        line = line.strip().rstrip(';')
+        if not line:
+            continue
+
+        part_params = line.split(':')
+        flags = [f for f in part_params[6].split(', ') if f != '']
+
+        parts.append({
+            'num':    int(part_params[0]),
+            'begin':  parse_unit(part_params[1], unit),
+            'end':    parse_unit(part_params[2], unit),
+            'size':   parse_unit(part_params[3], unit),
+            'fstype': part_params[4] or None,
+            'flags':  flags,
+        })
+
+    return { 'generic': generic, 'partitions': parts }
+
+
+def format_disk_size(size_bytes, unit, sector_size):
+    """
+    Formats a size in bytes into a different unit, like parted does. It doesn't
+    manage CYL and CHS formats, though.
+    This function has been adapted from https://github.com/Distrotech/parted/blo
+    b/279d9d869ff472c52b9ec2e180d568f0c99e30b0/libparted/unit.c
+    """
+    unit = unit.lower()
+
+    # If unit is the empty string, it defaults to 'compact'.
+    if unit in ['', 'compact', 'cyl', 'chs']:
+       unit = 'b'
+       if size_bytes >= 10 * math.pow(1000, 4):
+           unit = 'tb'
+       elif size_bytes >= 10 * math.pow(1000, 3):
+           unit = 'gb'
+       elif size_bytes >= 10 * math.pow(1000, 2):
+           unit = 'mb'
+       elif size_bytes >= 10 * math.pow(1000, 1):
+           unit = 'kb'
+
+    if unit == 's':
+        scale = sector_size
+    elif unit == '%':
+        scale = size_bytes / 100
+    elif unit == 'b':
+        scale = 1
+    elif unit == 'kb':
+        scale = math.pow(1000, 1)
+    elif unit == 'mb':
+        scale = math.pow(1000, 2)
+    elif unit == 'gb':
+        scale = math.pow(1000, 3)
+    elif unit == 'tb':
+        scale = math.pow(1000, 4)
+    elif unit == 'kib':
+        scale = math.pow(1024, 1)
+    elif unit == 'mib':
+        scale = math.pow(1024, 2)
+    elif unit == 'gib':
+        scale = math.pow(1024, 3)
+    elif unit == 'tib':
+        scale = math.pow(1024, 4)
```

Comment:
Both are fine IMO, there's nothing wrong with doing presence-checks either.
I'll let you decide.